File: Middle_School_Management_App/Middle_School_Management_App/EventController.swift
//
//  EventController.swift
//  Middle_School_Management_App
//
//  Created by Vivianne Yee on 2023-11-12.
//

import Foundation

class EventController {

    init() {}
    
    enum NetworkError: Error {
        case emptyResponse
        case invalidResponse
        case serverError(String)
    }
    
    // create a new event
    func createEvent(title: String, description: String, startDate: Date, endDate: Date, completion: @escaping (Result<GetEventResponse, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/event/events")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ" // Use the appropriate date format
        
        let body: [String: Any] = [
            "title": title,
            "description": description,
            "startDate": formatter.string(from: startDate),
            "endDate": formatter.string(from: endDate)
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body)
            request.httpBody = jsonData
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not create event")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 201 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let eventObject = try decoder.decode(GetEventResponse.self, from: responseData)
                completion(.success(eventObject))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    // get an event by its id
    func getEventById(id: String, completion: @escaping (Result<GetEventResponse, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/event/events/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not retrieve event")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let eventObject = try decoder.decode(GetEventResponse.self, from: responseData)
                completion(.success(eventObject))
            } catch {
                completion(.failure(error))
            }
                        
        }.resume()
    }
    
    // update an event
    func updateEvent(id: String, title: String, description: String, startDate: Date, endDate: Date, completion: @escaping (Result<GetEventResponse, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/event/events/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ" // Use the appropriate date format
        
        let body: [String: Any] = [
            "title": title,
            "description": description,
            "startDate": formatter.string(from: startDate),
            "endDate": formatter.string(from: endDate)
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body)
            request.httpBody = jsonData
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not update event")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let eventObject = try decoder.decode(GetEventResponse.self, from: responseData)
                completion(.success(eventObject))
            } catch {
                completion(.failure(error))
            }
            
        }.resume()
    }
    
    // delete an event
    func deleteEvent(id: String, completion: @escaping (Result<Data, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/event/events/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not delete event")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            completion(.success(responseData))
            
        }.resume()
    }
}

File: Backend/controllers/eventController.js
const Event = require('../models/Event')

// create a new event
exports.createEvent = async (req, res) => {

    const { title, description, startDate, endDate } = req.body

    try {
        // create new event
        const newEvent = new Event({
            title: title,
            description: description, 
            datePosted: Date(),
            startDate: startDate,
            endDate: endDate
        })

        await newEvent.save()

        return res.status(201).json({ message: 'Event created successfully' })
    } catch (error) {
        // catch server error
        console.error(error)
        return res.status(500).json({ message: 'Internal Server Error' })
    }
}

// get event by id
exports.getEventById = async (req, res) => {
    const eventId = req.params.id

    try {
        const retrievedEvent = await Event.findById(eventId)

        if (!retrievedEvent) {
            // 404 if event with eventId does not exist
            return res.status(404),json({ error: 'Event not found' })
        } else {
            // return 200 and event if successful
            return res.status(200).json({ message: 'Event retrieved successfully', 
                _id: retrievedEvent._id,
                title: retrievedEvent.title,
                datePosted: retrievedEvent.datePosted,
                startDate: retrievedEvent.startDate,
                endDate: retrievedEvent.endDate
            })
        }
    } catch (error) {
        // catch server error
        console.error(error)
        return res.status(500).json({ error: 'Internal Server Error' })
    }

}

// update event
exports.updateEvent = async (req, res) => {
    const eventId = req.params.id
    const { title, description, startDate, endDate} = req.body

    try {
        // find the event by id and update fields
        const updatedEvent = await Event.findByIdAndUpdate(eventId, { title: title, description: description, startDate: startDate, endDate: endDate }, { new: true })

        if (!updatedEvent) {
            // 404 if event with given id does not exist
            return res.status(404).json({ error: 'Event not found' })
        }

        // on success return 200 status, success message, and updated event
        return res.status(200).json({ message: 'Event retrieved successfully', 
            _id: retrievedEvent._id,
            title: retrievedEvent.title,
            datePosted: retrievedEvent.datePosted,
            startDate: retrievedEvent.startDate,
            endDate: retrievedEvent.endDate
        })    
    } catch (error) {
        // catch server error
        console.error(error)
        res.status(500).json({ error: 'Internal Server Error' })
    }
}

// delete event
exports.deleteEvent = async (req, res) => {
    const eventId = req.params.id

    try {
        // find the event by id and delete
        const deletedEvent = await Event.findByIdAndDelete(eventId)

        if (!deletedEvent) {
            // 404 if event with given id does not exist
            return res.status(404).json({ error: 'Event not found' })
        }

        res.status(200).json({ message: 'Event deleted successfully' })
    } catch (error) {
        // catch server error
        console.error(error)
        res.status(500).json({ error: 'Internal Server Error' })
    }
}